File: Design Strategies/Brute_Force.java
import java.util.*;

public class Brute_Force {
    static class Item {
        String name;
        int size;
        int value;

        Item(String name, int size, int value) {
            this.name = name;
            this.size = size;
            this.value = value;
        }
    }

    // Brute Force solution for 0/1 Knapsack Problem
    public static int knapsack(Item[] items, int capacity, List<Item> selected) {
        return knapsackHelper(items, capacity, 0, selected, new ArrayList<>());
    }

    private static int knapsackHelper(Item[] items, int capacity, int index, List<Item> selected, List<Item> current) {
        if (index == items.length || capacity == 0) {
            int value = current.stream().mapToInt(i -> i.value).sum();
            if (value > selected.stream().mapToInt(i -> i.value).sum()) {
                selected.clear();
                selected.addAll(current);
            }
            return value;
        }

        // Exclude current item
        int exclude = knapsackHelper(items, capacity, index + 1, selected, current);

        // Include current item if it fits
        int include = 0;
        if (items[index].size <= capacity) {
            current.add(items[index]);
            include = knapsackHelper(items, capacity - items[index].size, index + 1, selected, current);
            current.remove(current.size() - 1);
        }

        return Math.max(include, exclude);
    }

    public static void main(String[] args) {
        Item[] items = {
                new Item("A", 10, 60),
                new Item("B", 20, 100),
                new Item("C", 30, 120),
                new Item("D", 5, 50)
        };
        int capacity = 20;
        List<Item> selected = new ArrayList<>();
        int maxValue = knapsack(items, capacity, selected);

        System.out.println("Selected items:");
        for (Item item : selected) {
            System.out.println(item.name + " (size: " + item.size + ", value: " + item.value + ")");
        }
        System.out.println("Total value: " + maxValue);
    }
}

File: Design Strategies/Brute_Force.c
#include <stdio.h>
#include <string.h>

typedef struct {
    char name[20];
    int size;
    int value;
} Item;

int maxValue = 0;
int bestSelected[100];
int tempSelected[100];

int knapsack(Item items[], int n, int capacity, int index, int currValue, int currSize) {
    if (index == n || capacity == 0) {
        if (currValue > maxValue) {
            maxValue = currValue;
            memcpy(bestSelected, tempSelected, sizeof(int) * n);
        }
        return currValue;
    }

    // Exclude current item
    tempSelected[index] = 0;
    int exclude = knapsack(items, n, capacity, index + 1, currValue, currSize);

    // Include current item if it fits
    int include = 0;
    if (items[index].size <= capacity) {
        tempSelected[index] = 1;
        include = knapsack(items, n, capacity - items[index].size, index + 1, currValue + items[index].value, currSize + items[index].size);
    }

    return (include > exclude) ? include : exclude;
}

int main() {
    Item items[] = {
        {"A", 10, 60},
        {"B", 20, 100},
        {"C", 30, 120},
        {"D", 5, 50}
    };
    int n = sizeof(items) / sizeof(items[0]);
    int capacity = 20;

    memset(bestSelected, 0, sizeof(bestSelected));
    memset(tempSelected, 0, sizeof(tempSelected));
    int result = knapsack(items, n, capacity, 0, 0, 0);

    printf("Selected items:\n");
    for (int i = 0; i < n; i++) {
        if (bestSelected[i]) {
            printf("%s (size: %d, value: %d)\n", items[i].name, items[i].size, items[i].value);
        }
    }
    printf("Total value: %d\n", maxValue);

    return 0;
}